### PR TITLE
Proposed classes for GFlow.Node and GFlow.SimpleNode

### PR DESCRIPTION
**Pull request by esodan**
This classes are added but no build is done jet, are for review.

I've created a GFlow.Node interface and a GFlow.SimpleNode implementation, this is not GUI aware, but try to move away from GUI logic and provide an interface to be used by a GtkFlow.Node to be shown on screen. This is what I mean about.

### CODE REVIEW

**Comment by esodan** [comment]
GtkFlow.Node will should have a GFlow.Node property, with the information in this node GtkFlow.Node should show widgets to add, remove, and connect Dock objects.

Dock should be defined as an interface at GFlow, implemented it for Sink and Source classes. I would like to define Sink and Source as interfaces too, to be implemented by any other objects out there. Then provide SimpleSink and SimpleSource implementations to be used in GtkFlow.Source and GtkFlow.Sink widgets to draw them on screen.

GtkFlow should use just interfaces references, then from they API take the information it's widgets requires to draw them on screen.

**Comment by esodan** [comment]
Please consider that in C/GObject/Vala we can use virtual methods, they can use properties in the interface to operate on them. I'm not sure, but some code to remove_sink could be implemented by Dock interface to avoid it on implementations, but requires to define abstract properties as Public API, then this is when is important to know what type of collections to use, for this stage I recomend to use ones you are using now.

File: libgflow/gflow-node.vala [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+        public signal void sinks_changed ();
+        public signal void sources_changed ();
+        public abstract string name { get; set; }
+        public abstract void disconnect_all ();
+        public abstract bool is_recursive (Node from, bool initial=false);
+        public abstract Dock? get_dock (string name);
+        public abstract bool has_dock(Dock d);
+        public abstract unowned List<Source> get_sources ();
+        public abstract add_source (Source source);
+        public abstract remove_source (Source source);
```
Comment (grindhold) [inline libgflow/gflow-node.vala]: i've thought a bit about the methods add_source remove_source add_sink and remove_sink. actually we could replace them by add_dock and remove_dock and let the implementations decide how they handle the different types of docks.

**Comment by grindhold** [comment]
All in all this looks very good to me. Aside from the thing i mentioned in the line-comment. I like the new sinks_changed/sources_changed signals. If you wouldn't mind integrating the process of building two separate libraries, typelibs e.t.c into the autotools buildsystem, we're set to go.

**Comment by esodan** [comment]
I'm preparing all classes and interfaces here, before merge. You can comment any thing when necessary.

I'll setup GFlow library, when finish and it is working, you can merge please. Next step will be its integration with GtkFlow.

When GFlow is done, is necessary to setup Unit Tests for it.l

File: libgflow/gflow-node.vala (hunk 1 above)
Comment (esodan) [inline libgflow/gflow-node.vala]: As for Dock, could be a Source or Sink, then it should not have methods to add a Source or Sink objects.

A Source should provide Data as Fixed or Stream, if the first you can know its type, if later then could be a shunk of data.

A Sink should have a weak reference to Source objects.

A Node should have both lists of Sink and Source objects.

Am I understanding correctly?

**Comment by esodan** [comment]
Could you review my code in order to continue on Unit Tests please?

**Comment by grindhold** [comment]
looks fine :) i tested it on the existing examples. works nicely

**Pull request merged**